.crawler/src/crawler.rs

```diff
@@ -0,0 +1,274 @@
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{constants::*, known_network::KnownNetwork};
+use snarkos_environment::CurrentNetwork;
+use snarkos_network::Data;
+use snarkos_storage::BlockLocators;
+use snarkos_synthetic_node::{ClientMessage, SynthNode, MAXIMUM_FORK_DEPTH, MESSAGE_LENGTH_PREFIX_SIZE, MESSAGE_VERSION};
+use snarkvm::traits::Network;
+
+use pea2pea::{
+    protocols::{Disconnect, Handshake, Reading, Writing},
+    Config,
+    Node as Pea2PeaNode,
+    Pea2Pea,
+};
+use std::{convert::TryInto, io, net::SocketAddr, ops::Deref, sync::Arc, time::Duration};
+use structopt::StructOpt;
+use tokio::task;
+use tracing::*;
+
+// CLI
+#[derive(Debug, StructOpt)]
+pub struct Opts {
+    /// Specify the IP address and port for the node server.
+    /// Naming and defaults kept consistent with snarkOS.
+    #[structopt(parse(try_from_str), default_value = "0.0.0.0:4132", long = "node")]
+    pub node: SocketAddr,
+}
+
+#[derive(Clone)]
+pub struct Crawler {
+    synth_node: SynthNode,
+    known_network: Arc<KnownNetwork>,
+}
+
+impl Pea2Pea for Crawler {
+    fn node(&self) -> &Pea2PeaNode {
+        &self.synth_node.node()
+    }
+}
+
+impl Deref for Crawler {
+    type Target = SynthNode;
+
+    fn deref(&self) -> &Self::Target {
+        &self.synth_node
+    }
+}
+
+impl Crawler {
+    /// Creates a crawler node with the most basic network protocols enabled.
+    pub async fn new(opts: Opts) -> Self {
+        let config = Config {
+            listener_ip: Some(opts.node.ip()),
+            desired_listening_port: Some(opts.node.port()),
+            max_connections: MAXIMUM_NUMBER_OF_PEERS as u16,
+            ..Default::default()
+        };
+
+        let pea2pea_node = Pea2PeaNode::new(Some(config)).await.unwrap();
+        let client_state = Default::default();
+        let node = Self {
+            synth_node: SynthNode::new(pea2pea_node, client_state),
+            known_network: Arc::new(KnownNetwork::default()),
+        };
+
+        node.enable_disconnect().await;
+        node.enable_handshake().await;
+        node.enable_reading().await;
+        node.enable_writing().await;
+
+        node
+    }
+
+    /// Spawns a task dedicated to broadcasting Ping messages.
+    pub fn send_pings(&self) {
+        let node = self.clone();
+        task::spawn(async move {
+            let genesis = CurrentNetwork::genesis_block();
+            let ping_msg = ClientMessage::Ping(
+                MESSAGE_VERSION,
+                MAXIMUM_FORK_DEPTH,
+                node.node_type(),
+                node.state(),
+                genesis.hash(),
```

Comment:
not an immediate concern, but we might want to track the network's top height and use that one here instead so as not to suggest that there's a node out there that can't sync properly :smiley:.

Reply:
on second thought, that would require us to send proper block locators in `Pong`s :thinking:; let's leave it be, then, at least for now